### PR TITLE
[FXML-1893] Implement folding for additions of constant tensors

### PR DESCRIPTION
**Pull request by TinaAMD**

### CODE REVIEW

File: mlir/include/mlir/Dialect/Tosa/Transforms/TosaFoldCommon.h [hunk 1]
```diff
@@ -40,10 +40,12 @@ DenseElementsAttr applyElementWise(
 /// Apply the given transformation function on the elements of the given
 /// tensors. If the input tensors do not match \p targetType, broadcasting is
 /// applied.
+template <class FirstValType, class SecondValType, class ResultType>
 DenseElementsAttr applyElementWise(
     const DenseElementsAttr &first, const DenseElementsAttr &second,
     TensorType targetType,
-    const std::function<APFloat(const APFloat &, const APFloat &)> &toApply);
+    const std::function<ResultType(const FirstValType &, const SecondValType &)>
```
Comment (mgehre-amd) [inline mlir/include/mlir/Dialect/Tosa/Transforms/TosaFoldCommon.h:47]: Are there binary ops where first and second operands have different types?

File: mlir/lib/Dialect/Tosa/Transforms/TosaFoldCommon.cpp [hunk 2]
```diff
@@ -205,6 +218,32 @@ OffsetType mlir::tosa::getBroadcastedOffset(DimensionType desiredShape,
...
+    return true;
+  }
+
+  // If this is the only use of one of the tensors, it will be replaced an no
+  // additional memory is required.
+  if (firstOp.hasOneUse() || secondOp.hasOneUse()) {
+    return true;
+  }
+
+  // Fold it both inputs are equal and those are the only uses. Don't fold
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFoldCommon.cpp:240]: Nice catch!

**Review by mgehre-amd — APPROVED** [review]
Looks great!

File: mlir/include/mlir/Dialect/Tosa/Transforms/TosaFoldCommon.h (hunk 1 above)
Comment (TinaAMD) [reply mlir/include/mlir/Dialect/Tosa/Transforms/TosaFoldCommon.h:47]: Not really, should I change it to `ElementType`?

Comment (mgehre-amd) [reply mlir/include/mlir/Dialect/Tosa/Transforms/TosaFoldCommon.h:47]: yes, I think that would be clearer

**Pull request merged**